_posts/2020-10-01-torch-network-from-scratch/torch_network_from_scratch.Rmd

```diff
@@ -0,0 +1,1076 @@
+
+    torch_tensor 
+     1  2  3
+     4  5  6
+     7  8  9
+    [ CPULongType{3,3} ]
+
+In higher dimensions especially, it can be easier to specify the type of tensor abstractly, as in: "give me a tensor of \<...\> of shape n1 x n2", where \<...\> could be "zeros"; or "ones"; or, say, "values drawn from a standard normal distribution":
+
+```{r}
+# a 3x3 tensor of standard-normally distributed values
+t <- torch_randn(3, 3)
+t
+
+# a 4x2x2 (3d) tensor of zeroes
+t <- torch_zeros(4, 2, 2)
+t
+
+```
+
+    torch_tensor 
+    -2.1563  1.7085  0.5245
+     0.8955 -0.6854  0.2418
+     0.4193 -0.7742 -1.0399
+    [ CPUFloatType{3,3} ]
+
+    torch_tensor 
+    (1,.,.) = 
+      0  0
+      0  0
+
+    (2,.,.) = 
+      0  0
+      0  0
+
+    (3,.,.) = 
+      0  0
+      0  0
+
+    (4,.,.) = 
+      0  0
+      0  0
+    [ CPUFloatType{4,2,2} ]
+
+Many similar functions exist, including, e.g., `torch_arange()` to create a tensor holding a sequence of evenly spaced values, `torch_eye()` that returns an identity matrix, and `torch_logspace()` that will fill a specified range with a list of values spaced logarithmically.
+
+If no `dtype` argument is specified, `torch` will infer the data type from the passed-in value(s). For example:
+
+```{r}
+t <- torch_tensor(c(3, 5, 7))
+t$dtype
+
+t <- torch_tensor(1L)
+t$dtype
+```
+
+    torch_Float
+    torch_Long
+
+But we can explicitly request a different `dtype` if we want:
+
+```{r}
+t <- torch_tensor(2, dtype = torch_double())
+t$dtype
+```
+
+    torch_Double
+
+`torch` tensors live on a *device*. By default, this will be the CPU:
+
+```{r}
+t$device
+```
+
+    torch_device(type='cpu')
+
+But we could also define a tensor to live on the GPU:
+
+```{r}
+t <- torch_tensor(2, device = "cuda")
+t$device
+```
+
+    torch_device(type='cuda', index=0)
+
+We'll talk more about devices below.
+
+There is another very important parameter to the tensor creation functions: `requires_grad`. Here though, I need to ask for your patience: This one will prominently figure in the follow-up post.
+
+### Conversion to built-in R data types
+
+To convert `torch` tensors to R, use `as_array`:
+
+```{r}
+t <- torch_tensor(matrix(1:9, ncol = 3, byrow = TRUE))
+as_array(t)
+```
+
+         [,1] [,2] [,3]
+    [1,]    1    2    3
+    [2,]    4    5    6
+    [3,]    7    8    9
+
+Depending on whether the tensor is one-, two-, or three-dimensional, the resulting R object will be a vector, a matrix, or an array:
+
+```{r}
+t <- torch_tensor(c(1, 2, 3))
+as_array(t) %>% class()
+
+t <- torch_ones(c(2, 2))
+as_array(t) %>% class()
+
+t <- torch_ones(c(2, 2, 2))
+as_array(t) %>% class()
+```
+
+    [1] "numeric"
+
+    [1] "matrix" "array" 
+
+    [1] "array"
+
+For one-dimensional and two-dimensional tensors, it is also possible to use `as.integer` / `as.matrix`. (One reason you might want to do this is to have more self-documenting code.)
+
+If a tensor currently lives on the GPU, you need to move it to the CPU first:
```

Comment:
In order to work with it? If so, I'd go ahead and say that explicitly (if not here, then when you bring up "living" on the GPU earlier).

Reply:
I know you get to this later (now that I've read more), but it still might be nice to mention in the earlier section.